compiler/src/codegen/emitmod.re

```diff
@@ -1,8 +1,72 @@
 open Grain_typed;
 open Grain_utils;
-open Compmod;
+open Mashtree;
 
-let emit_module = ({asm, signature}, outfile) => {
+type bad_object =
+  | InvalidMagic
+  | InvalidVersion
+  | Corrupted;
+
+exception BadObject(bad_object);
+
+let emit_object = (mashed: mash_program, outfile) => {
+  Grain_utils.Fs_access.ensure_parent_directory_exists(outfile);
+  if (Config.debug^) {
+    let mash_string =
+      Sexplib.Sexp.to_string_hum @@ Mashtree.sexp_of_mash_program(mashed);
+    let mashfile = Filepath.String.replace_extension(outfile, "mashtree");
+    let oc = open_out(mashfile);
+    output_string(oc, mash_string);
+    close_out(oc);
+  };
+
+  let oc = open_out_bin(outfile);
+
+  output_string(oc, Cmi_format.magic);
+  let version_length = String.length(Config.version);
+  output_binary_int(oc, version_length);
+  output_string(oc, Config.version);
+  let cmi = Marshal.to_bytes(mashed.signature, []);
+  output_binary_int(oc, Bytes.length(cmi) + version_length + 12);
+  output_bytes(oc, cmi);
+  Marshal.to_channel(oc, mashed, []);
+
+  close_out(oc);
+};
+
+let load_object = (ic): mash_program => {
+  let read_magic = really_input_string(ic, 4);
+  if (read_magic != Cmi_format.magic) {
+    raise(BadObject(InvalidMagic));
+  };
+  let version_length = input_binary_int(ic);
+  let read_version = really_input_string(ic, version_length);
+  if (read_version != Config.version) {
```

Comment:
This current logic treats every version as breaking doesn't it? Which would require a full recompile even for small stdlib bug changes right?